arch: arm: implement TZ api and configure SAU and  non-secure core registers from secure firmware

ioannisg opened the pull request:
Define and implement an API to configure the Non-Secure instances
of core registers in ARM Cortex-M23 and Cortex-M33.

**Update**: This PR will, now, addresses most sub-tasks in #6766 

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: configure non-secure registers from secure firmware

Define and implement an API to configure the Non-Secure instances
of core registers in ARM Cortex-M23 and Cortex-M33.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
@andrewboie @galak @SebastianBoe 

Pushing this for feedback:

* I want to have an internal folder under arm/core/cortex_ that will implement all internal TrustZone API needed for ARM.
* An "include" header (currently tz.h) will  expose the required API.
* Source  folder will be included if `ARM_SECURE_FIRMWARE `is activated.

As an example, I implemented the first function which configures the Non-Secure  instances of core registers, so that a secure image can safely transit into non-secure state.

SebastianBoe commented on arch/arm/core/cortex_m/tz/arm_core_tz.c:
```diff
@@ -0,0 +1,47 @@
+ */
+
+#include <arch/arm/cortex_m/cmsis.h>
+#include "cortex_m/tz.h"
```
Should only use \" includes for relative paths.

ioannisg replied:
Thanks, will fix.

SebastianBoe commented on arch/arm/include/cortex_m/tz.h:
```diff
@@ -0,0 +1,79 @@
+
+#else
+
+#if defined(CONFIG_ARM_SECURE_FIRMWARE)
```
I haven't seen this before.

Could you add a comment explaining why this ifdef is necessary?

ioannisg replied:
Will do. 
In principle, any TrustZone -related implementation concerns a Secure Firmware (i.e. a firmware that is meant to run in secure state)

SebastianBoe commented on arch/arm/core/cortex_m/tz/arm_core_tz.c:
```diff
@@ -0,0 +1,47 @@
+	__TZ_set_PSP_NS(psp_ns);
+}
+
+static inline void configure_nonsecure_control(int spsel_ns,  int npriv_ns)
```
inline can be dropped.

The compiler will inline it if appropriate.

ioannisg replied:
ok

SebastianBoe commented on arch/arm/core/cortex_m/tz/arm_core_tz.c:
```diff
@@ -0,0 +1,47 @@
+	__TZ_set_PSP_NS(psp_ns);
+}
+
+static inline void configure_nonsecure_control(int spsel_ns,  int npriv_ns)
```
s/int spsel_ns/u32_t spsel_ns/

?

SebastianBoe commented:
"Is this a good structural design? Agree with naming of source folder?"

No comment.

ioannisg commented:
@SebastianBoe Implemented requested changes. Thanks for reviewing.

ioannisg committed:
arch: arm: internal API to enable/disable SAU

This commit defines and implements the internal ARM TrustZone
API to enable or disable the Security Attribution Unit.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: api for reading MCU SAU regions

Define and implement a function to read the number of
SAU regions configued for the MCU.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: configure NS stack pointer limit registers

This commit adds and implements an ARM internal API function
to configure the Non-Secure instance of MSPLIM or PSPLIM.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: API to configure fault target state

This commit defines and implements an internal ARMv8-M TrustZone
API to allow the user to configure the target state of fault
exceptions that are not banked between security states.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: API to configure Secure exception boost

This commit defines and implements an internal ARMv8-M TrustZone
API to allow the user to select whether Secure exceptions will
have priority boosting over Non-Secure exceptions.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

agross-linaro commented:
So if I understand this correctly, this is to configure non-secure from a secure zephyr image.  So are we running zephyr on both sides?

ioannisg commented:
@agross-linaro i believe this will, indeed, be the case is certain scenarios, when the secure image, besides being a secure bootloader, may serve as a more generic (background) image providing secure services to non secure world.

ioannisg committed:
arch: arm: include GCC -mcmse compile option for secure firmware

Make GCC compile with the -mcmse compile option, if we are
building a Secure firmware. The option will make Security
Extensions for secure executables available, and will set
the corresponding compile-time indicator flag, accordingly:
(i.e. __ARM_FEATURE_CMSE=3).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: API macros for CMSE non-secure function pointers

This commit defines and implements and internal ARM macro API
that allows the TrustZone user to declare, define, and evaluate
pointers of non-secure function type.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: macro API for defining non-secure entry functions

This commit introduces an internal ARM macro API to allow the
TrustZone user to define Non-Secure entry functions.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
@ruuddw FYI

ioannisg committed:
arch: arm: add option to define an NSC region

This commit adds K-config options that allow the user to
signify an ARM Secure Firmware that contains Secure Entry
functions and to define the starting address of the linker
section that will contain the Secure Entry functions. It
also instructs the linker to append the NSC section if
instructed so by the user.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: export entry veneers

When building a Secure firmware image, instruct the linker to
generate a symbol table with the entry veneers, to allow a Non
Secure firmware image to access Secure Entry functions via the
entry veneers.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

carlescufi commented:
@galak @agross-linaro @ioannisg any progress on this one? It's been sitting idle for a while now

ioannisg committed:
arch: arm: api to permit/block NS system reset requests

This commit contributes a simple API to allow the TrustZone
user to block or permit Non-Secure System Reset requests.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: API & implementation of SAU region configuration

This commit contribute an internal ARM API that enables the
user to confgure an ARM Security Attribution Unit region. It
also defines ARM_SAU as a K-config option to indicate that
an MCU implements the ARM SAU.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
@carlescufi (and all) This PR could, now, be reviewed as a complete patch (I removed the DNM flag).

SebastianBoe commented on arch/arm/core/cortex_m/tz/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+# '--section-start,.gnu.sgstubs' enables entry veneers (for Secure Entry
+# functions) to be placed inside the special output section .gnu.sgstubs.
+zephyr_ld_option_ifdef(CONFIG_ARM_FIRMWARE_HAS_SECURE_ENTRY_FUNCS
+	${LINKERFLAGPREFIX},--section-start,.gnu.sgstubs=${CONFIG_ARM_NSC_REGION_BASE_ADDRESS}
```
"All SG veneers are placed in the special output section .gnu.sgstubs. Its start address must be set, either with the command line option ‘--section-start’ or in a linker script, to indicate where to place these veneers in memory."

In general, it is better to do things in the linker script than in the build system.

If it was attempted and then deemed inappropriate to modify the linker script then documentation
for why we are deviating from this best practice should be documented here.

ioannisg replied:
@SebastianBoe, I followed another set of "best practices", been suggested by you, actually :) that is:
- not to pollute core ARM files (such as the arm linker.ld),
- generate as few new files as possible

So, with this approach we satisfy both principles, that is, we don't modify the core ARM linker script, and we don't generate a new file (e.g. a linker-script extension for TrustZone veneers). CMakeLists.txt is, anyway, necessary for all other build options, defined here.
This approach is also enabling us gather all build-time configuration for TrustZone in a single cmake file, which seems good to me.

agross-linaro replied:
There may be constraints on the region that a user may not know unless they are very familiar with the data sheets.  I am a little unsure about allowing a user to just specify some specific address.  And the default is definitely not right.

SebastianBoe reviewed (changes requested):
This needs to be covered by CI or else it will bitrot into vaporware.

ioannisg commented on arch/arm/core/cortex_m/tz/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+	${LINKERFLAGPREFIX},--out-implib=${CONFIG_ARM_ENTRY_VENEERS_LIB_NAME}
+)
+
+zephyr_ld_option_ifdef(CONFIG_ARM_FIRMWARE_HAS_SECURE_ENTRY_FUNCS
```
Yes, this is a good suggestion.

ioannisg replied:
fixed as suggested.

SebastianBoe commented on arch/arm/core/cortex_m/tz/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+# Security Extensions. This option is required when building a Secure firmware.
+zephyr_compile_options(-mcmse)
+
+if(${CONFIG_ARM_FIRMWARE_HAS_SECURE_ENTRY_FUNCS})
```
You should drop the ${}

ioannisg replied:
thanks, done.

SebastianBoe commented on arch/arm/core/cortex_m/tz/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+
+	zephyr_ld_options(
+		${LINKERFLAGPREFIX},--cmse-implib
+	)
```
Could you double-check that the indentation here is following the (de-facto) standard.

ioannisg replied:
Will fix this, thanks.

SebastianBoe reviewed (approved):
Build-system related changes are OK.

carlescufi commented:
@galak @agross-linaro any feedback on this one?

agross-linaro commented on arch/arm/include/cortex_m/tz.h:
```diff
@@ -66,6 +66,35 @@ typedef struct tz_nonsecure_setup_conf {
+ *
+ * @return N/A
+ */
+void tz_sau_configure(int enable, int allns);
```
nit on parameter name allns.  maybe all_ns.

ioannisg replied:
I think ARM names it SAU_CTRL.ALLNS, actually, but no problem with changing this :)

agross-linaro replied:
I'd take consistency with the ARM names over how it looks.

pizi-nordic commented on arch/arm/core/cortex_m/tz/arm_core_tz.c:
```diff
@@ -0,0 +1,149 @@
+	} else {
+		aircr_payload |= SCB_AIRCR_BFHFNMINS_Msk;
+	}
+	SCB->AIRCR = ((0x5FAUL << SCB_AIRCR_VECTKEY_Pos)
```
Could we use define(s) for magic numbers?

ioannisg replied:
Yes, I will make a define for it

pizi-nordic commented on arch/arm/core/cortex_m/tz/arm_core_tz.c:
```diff
@@ -0,0 +1,149 @@
+	} else {
+		aircr_payload &= ~(SCB_AIRCR_PRIS_Msk);
+	}
+	SCB->AIRCR = ((0x5FAUL << SCB_AIRCR_VECTKEY_Pos)
```
Could we use define(s) for magic numbers?

ioannisg replied:
Yes

pizi-nordic commented on arch/arm/include/cortex_m/tz.h:
```diff
@@ -0,0 +1,312 @@
+ *
+ * @param fptr supplied pointer to be checked
+ *
+ * @return non-zero if pointer can be of non-secure function type
```
Return value of 0 is OK in Zephyr:

```
if (XXX_CHECK()) {
 /* Exception */
  return -ECODE;
}
/* OK */
```

ioannisg replied:
Well, that is actually for function operations, where 0 indicates success, negative error codes indicate failures.
This macro is more of a boolean function, i.e. "is <something> <something else>?"

I will rename the define to:
`#define TZ_NONSECURE_FUNC_PTR_IS_NS(fptr) \` to make it more clear.

pizi-nordic replied:
TZ_NONSECURE_FUNC_PTR_IS_NS() looks OK for me.

pizi-nordic commented on arch/arm/core/cortex_m/tz/arm_core_tz.c:
```diff
@@ -0,0 +1,149 @@
+		TZ_SAU_Enable();
+	} else {
+		TZ_SAU_Disable();
+		if (allns) {
```
Is configuring disabled unit correct?

ioannisg replied:
I think so. When SAU is disabled we want to be able to choose whether the memory is by default NS or S.

pizi-nordic reviewed (commented):
The proposed API uses a lot of boolean flags as arguments, which make it hard to read. 

For example, without looking into the code/docs, you will be not able to check if `tz_sau_configure(1,1)` is correct and what it is really doing. IMHO it is better to write a bit more self-explanatory functions
 like "tz_sau_enable()", "tz_sau_set_all_nonsecure()" and so on.

agross-oss reviewed (commented):
Some comments.  I'd like to think more about the secure non-secure callable region placement.

carlescufi commented:
@ioannisg can you please reply to the comments from @pizi-nordic?
@agross-linaro any comments on this one?

ioannisg commented on arch/arm/core/cortex_m/tz/arm_core_tz.c:
```diff
@@ -0,0 +1,149 @@
+		__TZ_set_PSPLIM_NS(lim);
+	}
+	__ISB();
+	__DSB();
```
It seems that the instruction and data barriers are not needed when configuring NS instances from secure state, so I will remove them, thanks.

ioannisg commented on arch/arm/core/cortex_m/tz/arm_core_tz.c:
```diff
@@ -0,0 +1,149 @@
+	} else {
+		__TZ_set_PSPLIM_NS(lim);
+	}
+	__ISB();
```
Thans, will remove.

ioannisg committed:
arch: arm: add macro definition for vector key write permit value

This commit adds a macro definition for the Vector Key value of
AIRCR register that permits the accompanying writes. The macro
is then used instead of the hard-coded numerical value.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented on arch/arm/core/cortex_m/tz/arm_core_tz.c:
```diff
@@ -38,6 +38,19 @@ static void configure_nonsecure_control(u32_t spsel_ns, u32_t npriv_ns)
 }
 
+#if defined(CONFIG_ARMV8_M_MAINLINE)
+void tz_nonsecure_splim_set(u32_t is_msp, u32_t lim)
```
Finally, I decided to follow the suggestion and split the function into 2.

ioannisg commented on arch/arm/core/cortex_m/tz/arm_core_tz.c:
```diff
@@ -0,0 +1,149 @@
+	__TZ_set_PSP_NS(psp_ns);
+}
+
+static void configure_nonsecure_control(u32_t spsel_ns, u32_t npriv_ns)
```
I don't see how this could improve readability/ease-of-use for this function.
CONTROL register, clearly, has 2 bits widely used (SPSEL and nPRIV), so by having them as arguments, makes function easy to use, IMO.

pizi-nordic replied:
This will leave us with 2 boolean arguments, which are not so easy to read.
https://github.com/zephyrproject-rtos/zephyr/pull/6748#pullrequestreview-129070060

ioannisg replied:
Isn't this a private function? User should not be using it, AFAIK.

ioannisg replied:
Btw, the user is given a struct (properties) API to configure all Non-Secure instances of core registers.

ioannisg commented:
>The proposed API uses a lot of boolean flags as arguments, which make it hard to read.

>For example, without looking into the code/docs, you will be not able to check if tz_sau_configure(1,1) is correct and what it is really doing. IMHO it is better to write a bit more self-explanatory functions
like "tz_sau_enable()", "tz_sau_set_all_nonsecure()" and so on.

@pizi-nordic thanks for this feedback. In fact, I took this into consideration, but tried to have a balanced trade-off between ease-of-reading of the proposed API and total number of functions.
For the example you mentioned, I would need 3 functions, i.e.:
- enable_sau
-disable_sau_set_all_nonsecure
-disable_sau_set_all_secure

to have a fully-descriptive API. In  many cases I tried to follow your suggestion, but in some other cases, this would make us have too many API functions, which is also a factor that degrades API read-ability.
